xray/src/bin/merge_xray_quadtrees.rs

```diff
@@ -20,8 +27,179 @@ struct CommandlineArguments {
     input_directories: Vec<PathBuf>,
 }
 
+fn copy_images(input_directory: &Path, output_directory: &Path) {
+    if input_directory == output_directory {
+        return;
+    }
+    globwalk::GlobWalkerBuilder::new(
+        input_directory,
+        &format!("*.{}", xray::IMAGE_FILE_EXTENSION),
+    )
+    .build()
+    .expect("Failed to build GlobWalker")
+    .filter_map(Result::ok)
+    .for_each(|dir_entry| {
+        std::fs::copy(
+            dir_entry.path(),
+            output_directory.join(dir_entry.file_name()),
+        )
+        .expect("Failed to copy the file.");
+    })
+}
+
+fn copy_all_images(input_directories: &Vec<PathBuf>, output_directory: &Path) {
+    for input_directory in input_directories {
+        copy_images(input_directory, output_directory);
+    }
+}
+
+fn read_metadata(path: &Path) -> proto::Meta {
+    let data = std::fs::read(&path).expect("Cannot open file");
+    protobuf::parse_from_reader::<proto::Meta>(&mut Cursor::new(data))
+        .expect("Could not read meta proto")
+}
+
+fn read_metadata_from_directory(directory: &Path) -> Vec<proto::Meta> {
+    globwalk::GlobWalkerBuilder::new(directory, "meta*.pb")
+        .build()
+        .expect("Failed to build GlobWalker")
+        .filter_map(Result::ok)
+        .map(|dir_entry| read_metadata(dir_entry.path()))
+        .collect()
+}
+
+fn read_metadata_from_directories(directories: &Vec<PathBuf>) -> Vec<proto::Meta> {
+    directories
+        .iter()
+        .map(|directory| read_metadata_from_directory(&directory))
+        .flatten()
+        .collect()
+}
+
+fn get_root_node(meta: &proto::Meta) -> Option<NodeId> {
+    meta.get_nodes()
+        .iter()
+        .map(NodeId::from)
+        .min_by_key(|node| node.level())
+}
+
+fn get_root_nodes(meta: &Vec<proto::Meta>) -> Vec<Option<NodeId>> {
```

Comment:
Would it make sense to already return an `Option<Vec<NodeId>>` here? This can easily be done by just changing the return type, `collect()` will do the magic behind...
Then the logic inside `validate_metadata()` would become a little easier...

Reply:
Returning Vec<NodeId> while skipping metas with no nodes.